### PR TITLE
BUG: Fix 4DCT vrg pipeline not properly centering sequences

### LINKED ISSUE #105: Automatic camera placement for 4DCT
Hi, It looks like the automatic camera placement feature doesn't work for my data. It always the error below
image

### PR DESCRIPTION
**Pull request by NicerNewerCar**
Regression introduced in e036c2249dc564c2cca03914837248dae6d06666

Closes #105 

Also resolves this comment https://github.com/BrownBiomechanics/SlicerAutoscoperM/issues/102#issuecomment-2180735267

### CODE REVIEW

**Comment by NicerNewerCar** [comment]
- [ ] This for loop needs to be changed to `for i in range(1,numFrames)` so we don't repeat the first item in the sequence. 
https://github.com/BrownBiomechanics/SlicerAutoscoperM/blob/1e87e54d110b982335ff8493040351de083000d1/AutoscoperM/AutoscoperM.py#L570-L584

**Comment by jcfr** [comment]
We are getting close :pray: 

Few nitpicks:
* move fix into its own commit
* add the performance change into its own commit
* move the removal of the threadexecutor into its own pull request

File: AutoscoperM/AutoscoperM.py [hunk 1]
```diff
@@ -601,16 +613,20 @@ def onGenerateVRG(self):
...
+                volumeNode.HardenTransform()
+            elif tfmNode is not None:
+                tfmNode.Inverse()
+                for i in range(0, volumeNode.GetNumberOfDataNodes()):
+                    currentNode = volumeNode.GetNthDataNode(i)
+                    currentNode.SetAndObserveTransformNodeID(tfmNode.GetID())
+                    currentNode.HardenTransform()
+                    currentNode.SetAndObserveTransformNodeID(None)
+                    volumeNode.SetDataNodeAtValue(currentNode, f"{i}")
+            slicer.mrmlScene.RemoveNode(tfmNode)
```
Comment (jcfr) [inline AutoscoperM/AutoscoperM.py:627]: A comment should be added to explain why this is needed

**Review by jcfr — COMMENTED** [review]
in `onGenerateVRG`, after line 512, I suggest to simplify doing something like this:

```py
            # center the volume
            transformOutputDir = os.path.join(mainOutputDir, tfmPath)
            self.logic.createPathsIfNotExists(transformOutputDir)
            
            
            if not self.logic.IsSequenceVolume(volumeNode):
                IO.exportOrigin2DicomTransform(volumeNode, transformOutputDir)
            else:
                # Just export the first frame
                proxyNode = self.logic.getItemInSequence(volumeNode, 0)[0]
                IO.exportOrigin2DicomTransform(proxyNode, transformOutputDir)

                if tfmNode is not None:

                    # if still needed, explain here why calling "AddCenteringTransform" and reverting its effect is still needed
```


And here is the function `exportOrigin2DicomTransform`:


```
def exportOrigin2DicomTransform(volumeNode, exportDir):
    """If volume is not already centered, save corresponding transform as `Origin2Dicom.tfm`"""

    if volumeNode.IsCentered():
        return False

    volumeNode.AddCenteringTransform():

    # Get reference to the centering transform (Volume to Center)
    tfmNode = volumeNode.GetParentTransformNode()

    # Revert effect of "AddCenteringTransform"
    volumeNode.HardenTransform()
    volumeNode.SetAndObserveTransformNodeID(None)

    # Update from "Volume to Center" to "Center to Volume"
    tfmNode.Inverse()

    # Save as "Origin2Dicom.tfm"
    tfmPath = os.path.join(exportDir, "Origin2Dicom.tfm")
    slicer.util.saveNode(tfmNode, tfmPath)

    slicer.mrmlScene.RemoveNode(tfmNode)

    return True
```

Ideally, we should also look into adding support to call `GetCenterPositionRAS` from Python, this would simplify the logic.

**Comment by NicerNewerCar** [comment]
> Ideally, we should also look into adding support to call `GetCenterPositionRAS` from Python, this would simplify the logic.

This doesn't seem to be exposed in the Python API:
```python
>>> x.GetCenterPositionRAS()
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'MRMLCorePython.vtkMRMLScalarVolumeNode' object has no attribute 'GetCenterPositionRAS'
```

It also seems to be a [protected function](https://apidocs.slicer.org/master/classvtkMRMLVolumeNode.html#a986896fb9295134b86080f25b7590307) for the vtkMRMLVolumeNode. Was this exposed somewhere else?

**Comment by amymmorton** [comment]
Suggest with the [depreciation of vrgs](#137 )- also close this

**Pull request closed without merging**